### PR TITLE
Added missing With methods in CompilationOptions

### PR DESCRIPTION
**Pull request by keyboardDrummer**
Added With<memberName> methods in CompilationOptions that were still missing. Also added testcases to check whether the derived classes of CompilationOptions properly implement the abstract CommonWith<memberName> methods.

### CODE REVIEW

**Comment by jasonmalinowski** [comment]
@dotnet-bot test windows_release_unit32_prtest please and test windows_release_unit64_prtest please. (These are new jobs that are now also required.)

**Comment by jasonmalinowski** [comment]
Tagging @dotnet/roslyn-compiler for the review. @keyboardDrummer, you'll probably want to associate your GitHub account with the email address you're using in those commits, so it'll properly link up and give you credit in your profile page.

**Comment by jasonmalinowski** [comment]
Also @dotnet-bot test windows_release_unit32_prtest please and test windows_release_unit64_prtest please. (These are new jobs that are now also required.)

File: src/Compilers/Core/CodeAnalysisTest/CommonCompilationOptionsTests.cs [hunk 1]
```diff
@@ -1,6 +1,10 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
+using Microsoft.CodeAnalysis.CSharp;
 using Roslyn.Test.Utilities;
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
```
Comment (jcouv) [inline src/Compilers/Core/CodeAnalysisTest/CommonCompilationOptionsTests.cs]: Can those imports be reverted?

File: src/Compilers/Core/Portable/PublicAPI.Unshipped.txt [hunk 2]
```diff
@@ -14,6 +14,15 @@ Microsoft.CodeAnalysis.Compilation.CreateTupleTypeSymbol(System.Collections.Immu
...
+Microsoft.CodeAnalysis.CompilationOptions.WithCryptoKeyContainer(string cryptoKeyContainer) -> Microsoft.CodeAnalysis.CompilationOptions
+Microsoft.CodeAnalysis.CompilationOptions.WithCryptoKeyFile(string cryptoKeyFile) -> Microsoft.CodeAnalysis.CompilationOptions
+Microsoft.CodeAnalysis.CompilationOptions.WithCryptoPublicKey(System.Collections.Immutable.ImmutableArray<byte> cryptoPublicKey) -> Microsoft.CodeAnalysis.CompilationOptions
+Microsoft.CodeAnalysis.CompilationOptions.WithDelaySign(bool? delaySign) -> Microsoft.CodeAnalysis.CompilationOptions
+Microsoft.CodeAnalysis.CompilationOptions.WithMainTypeName(string mainTypeName) -> Microsoft.CodeAnalysis.CompilationOptions
+Microsoft.CodeAnalysis.CompilationOptions.WithModuleName(string moduleName) -> Microsoft.CodeAnalysis.CompilationOptions
+Microsoft.CodeAnalysis.CompilationOptions.WithOverflowChecks(bool checkOverflow) -> Microsoft.CodeAnalysis.CompilationOptions
+Microsoft.CodeAnalysis.CompilationOptions.WithScriptClassName(string scriptClassName) -> Microsoft.CodeAnalysis.CompilationOptions
+Microsoft.CodeAnalysis.CompilationOptions.WithWarningLevel(int warningLevel) -> Microsoft.CodeAnalysis.CompilationOptions
```
Comment (jcouv) [inline src/Compilers/Core/Portable/PublicAPI.Unshipped.txt]: Ping for @gafter @jaredpar since affecting public API.

**Comment by jcouv** [comment]
@dotnet-bot test windows_debug_unit64_prtest please

**Comment by jcouv** [comment]
LGTM Thanks

File: src/Compilers/Core/CodeAnalysisTest/CommonCompilationOptionsTests.cs (hunk 1 above)
Comment (keyboardDrummer) [inline src/Compilers/Core/CodeAnalysisTest/CommonCompilationOptionsTests.cs]: Done

File: src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt [hunk 3]
```diff
@@ -1,8 +1,8 @@
-*REMOVED*override sealed Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode.ToString() -> string
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt]: >*REMOVED* [](start = 0, length = 9)

It looks like there is no meaningful change in this file. Consider reverting.

File: src/Compilers/VisualBasic/Portable/PublicAPI.Unshipped.txt [hunk 4]
```diff
@@ -59,5 +62,3 @@ Shared Microsoft.CodeAnalysis.VisualBasic.SyntaxFactory.TupleExpression(openPare
...
 Shared Microsoft.CodeAnalysis.VisualBasic.SyntaxFactory.TupleType(openParenToken As Microsoft.CodeAnalysis.SyntaxToken, elements As Microsoft.CodeAnalysis.SeparatedSyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.TupleElementSyntax), closeParenToken As Microsoft.CodeAnalysis.SyntaxToken) -> Microsoft.CodeAnalysis.VisualBasic.Syntax.TupleTypeSyntax
-Microsoft.CodeAnalysis.VisualBasic.VisualBasicParseOptions.New(languageVersion As Microsoft.CodeAnalysis.VisualBasic.LanguageVersion = Microsoft.CodeAnalysis.VisualBasic.LanguageVersion.Default, documentationMode As Microsoft.CodeAnalysis.DocumentationMode = Microsoft.CodeAnalysis.DocumentationMode.Parse, kind As Microsoft.CodeAnalysis.SourceCodeKind = Microsoft.CodeAnalysis.SourceCodeKind.Regular, preprocessorSymbols As System.Collections.Generic.IEnumerable(Of System.Collections.Generic.KeyValuePair(Of String, Object)) = Nothing) -> Void
-Microsoft.CodeAnalysis.VisualBasic.VisualBasicParseOptions.SpecifiedLanguageVersion() -> Microsoft.CodeAnalysis.VisualBasic.LanguageVersion
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/PublicAPI.Unshipped.txt]: It looks like these two lines are duplicates of lines added above.

File: src/Compilers/VisualBasic/Portable/PublicAPI.Unshipped.txt [hunk 5]
```diff
@@ -35,6 +35,9 @@ Microsoft.CodeAnalysis.VisualBasic.Syntax.TupleTypeSyntax.WithOpenParenToken(ope
...
 Microsoft.CodeAnalysis.VisualBasic.SyntaxKind.TupleExpression = 788 -> Microsoft.CodeAnalysis.VisualBasic.SyntaxKind
 Microsoft.CodeAnalysis.VisualBasic.SyntaxKind.TupleType = 789 -> Microsoft.CodeAnalysis.VisualBasic.SyntaxKind
+Microsoft.CodeAnalysis.VisualBasic.VisualBasicCompilationOptions.WithWarningLevel(value As Integer) -> Microsoft.CodeAnalysis.VisualBasic.VisualBasicCompilationOptions
+Microsoft.CodeAnalysis.VisualBasic.VisualBasicParseOptions.New(languageVersion As Microsoft.CodeAnalysis.VisualBasic.LanguageVersion = Microsoft.CodeAnalysis.VisualBasic.LanguageVersion.Default, documentationMode As Microsoft.CodeAnalysis.DocumentationMode = Microsoft.CodeAnalysis.DocumentationMode.Parse, kind As Microsoft.CodeAnalysis.SourceCodeKind = Microsoft.CodeAnalysis.SourceCodeKind.Regular, preprocessorSymbols As System.Collections.Generic.IEnumerable(Of System.Collections.Generic.KeyValuePair(Of String, Object)) = Nothing) -> Void
+Microsoft.CodeAnalysis.VisualBasic.VisualBasicParseOptions.SpecifiedLanguageVersion() -> Microsoft.CodeAnalysis.VisualBasic.LanguageVersion
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/PublicAPI.Unshipped.txt]: There were no changes to ParseOptions. Why is this change needed?

File: src/Compilers/VisualBasic/Portable/VisualBasicCompilationOptions.vb [hunk 6]
```diff
@@ -1199,5 +1199,54 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
+        ''' </summary>
+        ''' <param name="value">The warning level setting. </param>        
+        ''' <returns>A new instance of VisualBasicCompilationOptions, if the warning level is different; otherwise current instance.</returns>        
+        Public Shadows Function WithWarningLevel(value As Integer) As VisualBasicCompilationOptions
+            If value = Me.WarningLevel Then
+                Return Me
+            End If
+
+            Return New VisualBasicCompilationOptions(Me) With {.WarningLevel = value}
+        End Function
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/VisualBasicCompilationOptions.vb]: We probably shouldn't add this member to VB, it is not one of the options affecting it. We probably shouldn't add it to the common options as well.

**Comment by mattwar** [comment]
Looks good.. but we need to find out why the solution crawler test is failing.

**Comment by mattwar** [comment]
@dotnet-bot test windows_debug_unit32_prtest pelase

**Comment by mattwar** [comment]
@dotnet-bot test windows_debug_unit32_prtest please

File: src/Compilers/VisualBasic/Test/Semantic/Compilation/VisualBasicCompilationOptionsTests.vb [hunk 7]
```diff
@@ -83,7 +124,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.UnitTests
             TestProperty(Function(old, value) old.WithSourceReferenceResolver(value), Function(opt) opt.SourceReferenceResolver, New SourceFileResolver(ImmutableArray(Of String).Empty, Nothing))
             TestProperty(Function(old, value) old.WithMetadataReferenceResolver(value), Function(opt) opt.MetadataReferenceResolver, New TestMetadataReferenceResolver())
             TestProperty(Function(old, value) old.WithAssemblyIdentityComparer(value), Function(opt) opt.AssemblyIdentityComparer, New DesktopAssemblyIdentityComparer(New AssemblyPortabilityPolicy()))
-            TestProperty(Function(old, value) old.WithStrongNameProvider(value), Function(opt) opt.StrongNameProvider, New DesktopStrongNameProvider())
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/Compilation/VisualBasicCompilationOptionsTests.vb]: >TestProperty(Function(old, value) old.WithStrongNameProvider(value), Function(opt) opt.StrongNameProvider, New DesktopStrongNameProvider()) [](start = 12, length = 139)

Why is this code removed?

**Comment by mattwar** [comment]
@dotnet-bot test windows_release_unit32_prtest please

**Comment by mattwar** [comment]
@keyboardDrummer can you respond to feedback from @AlekseyTs ?

File: src/Compilers/VisualBasic/Portable/PublicAPI.Unshipped.txt (hunk 4 above)
Comment (keyboardDrummer) [inline src/Compilers/VisualBasic/Portable/PublicAPI.Unshipped.txt]: Yes the automated refactoring that updated this file moved these lines. I'll undo the move.

File: src/Compilers/VisualBasic/Portable/PublicAPI.Unshipped.txt (hunk 5 above)
Comment (keyboardDrummer) [inline src/Compilers/VisualBasic/Portable/PublicAPI.Unshipped.txt]: The same answer as above: the automated refactoring that updated this file moved these lines. I'll undo the move.

File: src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt (hunk 3 above)
Comment (keyboardDrummer) [inline src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt]: Reverted. I should learn not to let the PublicAPI files update through the automatic code fix ^,^

File: src/Compilers/VisualBasic/Portable/VisualBasicCompilationOptions.vb (hunk 6 above)
Comment (keyboardDrummer) [inline src/Compilers/VisualBasic/Portable/VisualBasicCompilationOptions.vb]: You are right. I've removed it from both the VB and the common options.

File: src/Compilers/VisualBasic/Test/Semantic/Compilation/VisualBasicCompilationOptionsTests.vb (hunk 7 above)
Comment (keyboardDrummer) [inline src/Compilers/VisualBasic/Test/Semantic/Compilation/VisualBasicCompilationOptionsTests.vb]: Yikes, that was an accident. Thanks for spotting it. I've reverted the change.

File: src/Compilers/VisualBasic/Portable/PublicAPI.Unshipped.txt [hunk 8]
```diff
@@ -60,4 +60,4 @@ Shared Microsoft.CodeAnalysis.VisualBasic.SyntaxFactory.TupleType(ParamArray ele
...
 Microsoft.CodeAnalysis.VisualBasic.VisualBasicParseOptions.New(languageVersion As Microsoft.CodeAnalysis.VisualBasic.LanguageVersion = Microsoft.CodeAnalysis.VisualBasic.LanguageVersion.Default, documentationMode As Microsoft.CodeAnalysis.DocumentationMode = Microsoft.CodeAnalysis.DocumentationMode.Parse, kind As Microsoft.CodeAnalysis.SourceCodeKind = Microsoft.CodeAnalysis.SourceCodeKind.Regular, preprocessorSymbols As System.Collections.Generic.IEnumerable(Of System.Collections.Generic.KeyValuePair(Of String, Object)) = Nothing) -> Void
-Microsoft.CodeAnalysis.VisualBasic.VisualBasicParseOptions.SpecifiedLanguageVersion() -> Microsoft.CodeAnalysis.VisualBasic.LanguageVersion
\ No newline at end of file
+Microsoft.CodeAnalysis.VisualBasic.VisualBasicParseOptions.SpecifiedLanguageVersion() -> Microsoft.CodeAnalysis.VisualBasic.LanguageVersion
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/PublicAPI.Unshipped.txt]: It looks like there is no meaningful change in this file. Consider reverting.

**Comment by AlekseyTs** [comment]
LGTM, modulo unexpected change in src/Compilers/VisualBasic/Portable/PublicAPI.Unshipped.txt

File: src/Compilers/VisualBasic/Portable/PublicAPI.Unshipped.txt (hunk 8 above)
Comment (keyboardDrummer) [inline src/Compilers/VisualBasic/Portable/PublicAPI.Unshipped.txt]: Done

**Comment by AlekseyTs** [comment]
LGTM

**Comment by keyboardDrummer** [comment]
@dotnet-bot test windows_debug_unit32_prtest please and windows_debug_unit64_prtest and windows_eta_open_prtest please

**Comment by keyboardDrummer** [comment]
@dotnet-bot test windows_debug_unit32_prtest please and test windows_debug_unit64_prtest and test windows_eta_open_prtest please

**Comment by keyboardDrummer** [comment]
@dotnet-bot test windows_debug_unit64_prtest please
